Docs site (#121)
electric-sql/pglite: docs/docs/api.md

---
outline: [2, 3]
---

# PGlite API

## Main Constructor

`new PGlite(dataDir: string, options: PGliteOptions)`<br/>
`new PGlite(options: PGliteOptions)`

> A new PGlite instance is created using the `new PGlite()` constructor.

Reviewer: might be worth deciding on whether we want `create` as the main initialization before release?

This is imported as:

```ts
import { PGlite } from "@electric-sql/pglite";
```

`await PGlite.create(dataDir: string, options: PGliteOptions)`<br />
`await PGlite.create(options: PGliteOptions)`

There is also a `PGlite.create()` static method that returns a promise, resolving to the new PGlite instance. There are a couple of advantages to using the static method:

> - This awaits the [`.waitReady`](#waitready) promise, ensuring that the database has fully initiated.

Reviewer: ```suggestion
- This awaits the [`.waitReady`](#waitready) promise, ensuring that the database has been fully iniialised.
```

> - When using TypeScript and extensions, the returned PGlite instance will have the extensions namespace on its type. This is not possible with the standard constructor due to limitation with TypeScript.

Reviewer: ```suggestion
- When using TypeScript and extensions, the returned PGlite instance will have the extensions namespace on its type. This is not possible with the standard constructor due to TypesScript limitations.
```

if you have a thread/issue for the limitations perhaps we could link it here? Or not mention it at all if we only push the static factory method.

#### `dataDir`

> Path to the directory for storing the Postgres database. You can provide a url scheme for various storage backends:

Reviewer: ```suggestion
Path to the directory for storing the Postgres database. You can provide a URI scheme for various storage backends:
```

- `file://` or unprefixed<br />
  File system storage, available in Node and Bun.
- `idb://`<br />

>   IndexedDB storage, available in the browser.

Reviewer: ```suggestion
  [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) storage, available in the browser.
```

- `memory://`<br />
  In-memory ephemeral storage, available in all platforms.

#### `options`

- `dataDir: string`<br />

>   The directory to store the Postgres database in when not provided as the first argument.

Reviewer: ```suggestion
  The directory in which to store the Postgres database when not provided as the first argument.
```

- `debug: 1-5`<br />

>   the Postgres debug level. Logs are sent to the console.

Reviewer: ```suggestion
  The Postgres debug level. Logs are sent to the console.
```

- `relaxedDurability: boolean`<br />

>   Under relaxed durability mode, PGlite will not wait for flushes to storage to complete after each query before returning results. This is particularly useful when using the indexedDB file system.

Reviewer: ```suggestion
  Under relaxed durability mode, PGlite will not wait for flushes to storage to complete after each query before returning results. This is particularly useful when using the IndexedDB file system.
```

- `fs: Filesystem`<br />

>   The alternative to providing a dataDir with a filesystem prefix is to initiate the Filesystem yourself and provide it here. See [Filesystems](./filesystems.md)

Reviewer: ```suggestion
  The alternative to providing a dataDir with a filesystem prefix is to initialise a `Filesystem` yourself and provide it here. See [Filesystems](./filesystems.md)
```

- `loadDataDir: Blob | File`<br />
  A tarball of a PGlite `datadir` to load when the database starts. This should be a tarball produced from the related [`.dumpDataDir()`](#dumpdatadir) method.
- `extensions: Extensions`<br />
  An object containing the extensions you wish to load.

#### `options.extensions`

PGlite and Postgres extensions are loaded into a PGLite instance on start, and can include both a WASM build of a Postgres extension and/or a PGlite client plugin.

> The `options.extensions` parameter is an object of `namespace: extension` parings. The namespace if used to expose the PGlite client plugin included in the extension. An example of this is the [live queries](./live-queries.md) extension.

Reviewer: ```suggestion
The `options.extensions` parameter is an object of `namespace: extension` parings. The namespace is used to expose the PGlite client plugin included in the extension. An example of this is the [live queries](./live-queries.md) extension.
```

import { vector } from "@electric-sql/pglite/vector";

const pg = await PGlite.create({
  extensions: {
    live, // Live query extension, is a PGlite client plugin
    vector, // Postgres pgvector extension
  },
});

// The `live` namespace is added by the use of the 
// `live` key in the `extensions` object.
pg.live.query('...')
```

For information on how to develop a PGlite extension see [Extension Development](../extensions/development.md).

## Methods

### query

`.query<T>(query: string, params?: any[], options?: QueryOptions): Promise<Results<T>>`

Execute a single statement, optionally with parameters.

Uses the *extended query* Postgres wire protocol.

Returns single [result object](#results-objects).

##### Example

```ts
await pg.query(
  'INSERT INTO test (name) VALUES ($1);',
  [ 'test' ]
);
// { affectedRows: 1 },
```

##### Query Options

The `query` and `exec` methods take an optional `options` objects with the following parameters:

- `rowMode: "object" | "array"` <br />
  The returned row object type, either an object of `fieldName: value` mappings or an array of positional values. Defaults to `"object"`.
- `parsers: ParserOptions` <br />

>   An object of type  `{[[pgType: number]: (value: string) => any;]}` mapping Postgres data type id to parser function.

Reviewer: ```suggestion
  An object of type  `{[[pgType: number]: (value: string) => any;]}` mapping Postgres data type IDs to parser functions.
```

  For convenience, the `pglite` package exports a constant for most common Postgres types:

  ```ts
  import { types } from "@electric-sql/pglite";
  await pg.query(`
    SELECT * FROM test WHERE name = $1;
  `, ["test"], {
    rowMode: "array",
    parsers: {
      [types.TEXT]: (value) => value.toUpperCase(),
    }
  });
  ```

- `blob: Blob | File` <br />
  Attach a `Blob` or `File` object to the query that can used with a `COPY FROM` command by using the virtual `/dev/blob` device, see [importing and exporting](#importing-and-exporting-with-copy-tofrom).

### exec

`.exec(query: string, options?: QueryOptions): Promise<Array<Results>>`

Execute one or more statements. *(note that parameters are not supported)*

> This is useful for applying database migrations, or running multi-statement sql that doesn't use parameters.

Reviewer: ```suggestion
This is useful for applying database migrations, or running multi-statement SQL that doesn't use parameters.
```

### offNotification

`offNotification(callback: (channel: string, payload: string) => void): void`

Remove an event handler for all notifications received from Postgres.

### dumpDataDir

`dumpDataDir(): Promise<File | Blob>`

Dump the Postgres `datadir` to a Gzipped tarball.

This can then be used in combination with the [`loadDataDir`](#options) option when starting PGlite to load a dumped database from storage.

::: tip NOTE

The datadir dump may not be compatible with other Postgres versions; it is only designed for importing back into PGlite.

:::

## Properties

### ready

`.ready` *`boolean (read only)`*

Whether the database is ready to accept queries.

### closed

`.closed` *`boolean (read only)`*

Whether the database is closed and no longer accepting queries.

### waitReady

`.waitReady` *`Promise<void>`*

Promise that resolves when the database is ready to use.

::: tip NOTE

Query methods will wait for the `waitReady` promise to resolve if called before the database has fully initialised, and so it is not necessary to wait for it explicitly.

:::

## `Results<T>` Objects

Result objects have the following properties:

- `rows: Row<T>[]`<br />

>   The rows retuned by the query

Reviewer: ```suggestion
  The rows retuned by the query.
```

`offNotification(callback: (channel: string, payload: string) => void): void`

Remove an event handler for all notifications received from Postgres.

### dumpDataDir

`dumpDataDir(): Promise<File | Blob>`

Dump the Postgres `datadir` to a Gzipped tarball.

This can then be used in combination with the [`loadDataDir`](#options) option when starting PGlite to load a dumped database from storage.

::: tip NOTE

The datadir dump may not be compatible with other Postgres versions; it is only designed for importing back into PGlite.

:::

## Properties

### ready

`.ready` *`boolean (read only)`*

Whether the database is ready to accept queries.

### closed

`.closed` *`boolean (read only)`*

Whether the database is closed and no longer accepting queries.

### waitReady

`.waitReady` *`Promise<void>`*

Promise that resolves when the database is ready to use.

::: tip NOTE

Query methods will wait for the `waitReady` promise to resolve if called before the database has fully initialised, and so it is not necessary to wait for it explicitly.

:::

## `Results<T>` Objects

Result objects have the following properties:

- `rows: Row<T>[]`<br />
  The rows retuned by the query.

> - `affectedRows?: number` <br />

Author: Yep, optional using TS syntax
